### PR TITLE
querier: add prometheus/api/v1/format_query API

### LINKED ISSUE #4260: Support `api/v1/format_query` for compatibility with Prometheus
Hi, not sure if valid or not for Mimir.
I suppose formatting does Grafana also, but we have Prometheus UI running on top of Mimir and that supports server-side query formatting, but that is not working with Mimir right now.

Querier/query-frontend could implement this to be fully compatible with the Prometheus API

### PR DESCRIPTION
**Pull request by FUSAKLA**
Resolves #4260 

Seems to be working fine, 

```bash
curl '0.0.0.0:8080/prometheus/api/v1/format_query?query=count(up)by(foo)'
{"status":"success","data":"count by (foo) (up)"}
```

### CODE REVIEW

File: pkg/api/handlers.go [hunk 1]
```diff
@@ -291,6 +291,7 @@ func NewQuerierHandler(
 	seriesQueryStats := usagestats.NewRequestsMiddleware("querier_series_query_requests")
 	metadataQueryStats := usagestats.NewRequestsMiddleware("querier_metadata_query_requests")
 	cardinalityQueryStats := usagestats.NewRequestsMiddleware("querier_cardinality_query_requests")
+	otherStats := usagestats.NewRequestsMiddleware("querier_other_requests")
```
Comment (FUSAKLA) [inline pkg/api/handlers.go:294]: It felt like it doesn't fit to any other stats classes, but not sure if others is ok? #naming

File: CHANGELOG.md [hunk 2]
```diff
@@ -57,6 +57,7 @@ Querying with using `{__mimir_storage__="ephemeral"}` selector no longer works.
 * [ENHANCEMENT] Compactor: added the experimental flag `-compactor.block-upload.block-validation-enabled` with the default `true` to configure whether block validation occurs on backfilled blocks. #3411
 * [ENHANCEMENT] Ingester: apply a jitter to the first TSDB head compaction interval configured via `-blocks-storage.tsdb.head-compaction-interval`. Subsequent checks will happen at the configured interval. This should help to spread the TSDB head compaction among different ingesters over the configured interval. #4364
 * [ENHANCEMENT] Ingester: the maximum accepted value for `-blocks-storage.tsdb.head-compaction-interval` has been increased from 5m to 15m. #4364
+* [ENHANCEMENT] Query-fronted,Querier: Added new HTTP API endpoint `<prometheus-http-prefix>/api/v1/format_query`. #TODO
```
Comment (FUSAKLA) [inline CHANGELOG.md:60]:
```suggestion
* [ENHANCEMENT] Query-fronted,Querier: Added new HTTP API endpoint `<prometheus-http-prefix>/api/v1/format_query`. #4373
```

File: docs/sources/mimir/references/http-api/index.md [hunk 3]
```diff
@@ -238,6 +239,14 @@ GET <alertmanager-http-prefix>/api/v1/status/buildinfo
 
 This endpoint returns in JSON format information about the build and enabled features. The format returned is not identical, but is similar to the [Prometheus Build Information endpoint](https://prometheus.io/docs/prometheus/latest/querying/api/#build-information).
 
+### Format query
+
+```
+GET,POST <prometheus-http-prefix>/api/v1/format_query?query=up
+```
+
+Formats the PromQL query, see [Prometheus documentation](https://prometheus.io/docs/prometheus/latest/querying/api/#formatting-query-expressions).
```
Comment (charleskorn) [inline docs/sources/mimir/references/http-api/index.md:248]: Let's follow the same pattern as for the other API endpoints:

```suggestion
Formats the PromQL query.

This endpoint is compatible with the Prometheus format query endpoint.

For more information about formatting queries, refer to Prometheus [formatting query expreessions](https://prometheus.io/docs/prometheus/latest/querying/api/#formatting-query-expressions).
```

Comment (charleskorn) [reply docs/sources/mimir/references/http-api/index.md:248]: Also, does this endpoint require authentication like the others do? If so, we should mention that here.

File: pkg/api/handlers.go (hunk 1 above)
Comment (charleskorn) [reply pkg/api/handlers.go:294]: All of the other statistics instances are specific, so I think we can be specific here as well: something like `querier_formatting_requests` perhaps?

**Review by charleskorn — COMMENTED** [review]
Thanks for the PR @FUSAKLA. Could you please add a test to cover this change?

File: CHANGELOG.md [hunk 4]
```diff
@@ -57,6 +57,7 @@ Querying with using `{__mimir_storage__="ephemeral"}` selector no longer works.
 * [ENHANCEMENT] Compactor: added the experimental flag `-compactor.block-upload.block-validation-enabled` with the default `true` to configure whether block validation occurs on backfilled blocks. #3411
 * [ENHANCEMENT] Ingester: apply a jitter to the first TSDB head compaction interval configured via `-blocks-storage.tsdb.head-compaction-interval`. Subsequent checks will happen at the configured interval. This should help to spread the TSDB head compaction among different ingesters over the configured interval. #4364
 * [ENHANCEMENT] Ingester: the maximum accepted value for `-blocks-storage.tsdb.head-compaction-interval` has been increased from 5m to 15m. #4364
+* [ENHANCEMENT] Query-fronted,Querier: Added new HTTP API endpoint `<prometheus-http-prefix>/api/v1/format_query`. #4373
```
[the remark this answers is not in the document]
Comment (FUSAKLA) [reply CHANGELOG.md]: :+1: 61428be06

File: docs/sources/mimir/references/http-api/index.md [hunk 5]
```diff
@@ -238,6 +239,14 @@ GET <alertmanager-http-prefix>/api/v1/status/buildinfo
 
 This endpoint returns in JSON format information about the build and enabled features. The format returned is not identical, but is similar to the [Prometheus Build Information endpoint](https://prometheus.io/docs/prometheus/latest/querying/api/#build-information).
 
+### Format query
+
+```
+GET,POST <prometheus-http-prefix>/api/v1/format_query?query=up
```
[the remark this answers is not in the document]
Comment (FUSAKLA) [reply docs/sources/mimir/references/http-api/index.md]: 49ac31a68

File: pkg/api/handlers.go (hunk 1 above)
Comment (FUSAKLA) [reply pkg/api/handlers.go]: Ok, renamed as suggested 08178f8ba

**Comment by FUSAKLA** [comment]
Hi, thanks for the review @charleskorn ! 

All comments should be resolved now, except for the tests.
The other endpoints which are forwarded just to the `promRouter` are not tested, so I do not have what to follow, and I'm not sure  what exactly to test.

Suppose just the registration of the endpoint? Do you have any suggestions? 
I'd be happy for any ideas and happy to implement it.

File: docs/sources/mimir/references/http-api/index.md [hunk 6]
```diff
@@ -239,6 +240,19 @@ GET <alertmanager-http-prefix>/api/v1/status/buildinfo
...
+```
+GET <prometheus-http-prefix>/api/v1/format_query?query={query}
+POST <prometheus-http-prefix>/api/v1/format_query
+```
+
+Formats the PromQL query.
+
+This endpoint is compatible with the Prometheus format query endpoint.
+
+For more information about formatting queries, refer to Prometheus [formatting query expreessions](https://prometheus.io/docs/prometheus/latest/querying/api/#formatting-query-expressions).
```
Comment (charleskorn) [inline docs/sources/mimir/references/http-api/index.md:254]:
```suggestion
For more information about formatting queries, refer to [Prometheus' documentation](https://prometheus.io/docs/prometheus/latest/querying/api/#formatting-query-expressions).
```

**Review by charleskorn — COMMENTED** [review]
Thanks @FUSAKLA! 

> Suppose just the registration of the endpoint? Do you have any suggestions?
> I'd be happy for any ideas and happy to implement it.

Perhaps a simple integration test?

**Comment by FUSAKLA** [comment]
Oh, great, didn't see the integration tests before!

Should be tested now https://github.com/grafana/mimir/pull/4373/commits/e93bbef44e9763ad532779f9f9d5dff70f4e7c74

File: integration/api_endpoints_test.go [hunk 7]
```diff
@@ -83,3 +83,19 @@ func TestConfigAPIEndpoint(t *testing.T) {
...
+	// Start Mimir in single binary mode, reading the config from file
+	s, mimir1 := newMimirSingleBinaryWithLocalFilesytemBucket(t, "mimir-1", nil)
+	defer s.Close()
+	require.NoError(t, s.StartAndWaitReady(mimir1))
+
+	// Get config from /prometheus/api/v1/format_query API endpoint.
+	res, err := e2e.DoGet(fmt.Sprintf("http://%s/prometheus/api/v1/format_query?query=count(up)by(foo)", mimir1.Endpoint(9009)))
+	require.NoError(t, err)
+
+	defer runutil.ExhaustCloseWithErrCapture(&err, res.Body, "config API response")
```
Comment (charleskorn) [inline integration/api_endpoints_test.go:97]:
```suggestion
	defer runutil.ExhaustCloseWithErrCapture(&err, res.Body, "format query API response")
```

File: integration/api_endpoints_test.go [hunk 8]
```diff
@@ -83,3 +83,19 @@ func TestConfigAPIEndpoint(t *testing.T) {
...
+	_, err = io.ReadAll(res.Body)
```
Comment (charleskorn) [inline integration/api_endpoints_test.go:98]: Should we make an assertion on the contents of the response?

**Review by charleskorn — COMMENTED** [review]
Thanks for adding the test @FUSAKLA. I've got two comments on the test and then I think this is good to go 🚀

File: integration/api_endpoints_test.go (hunk 8 above)
Comment (FUSAKLA) [reply integration/api_endpoints_test.go]: right, done https://github.com/grafana/mimir/pull/4373/commits/58d017091f77e794d64bb1729e4b3ec23177836b

**Review by charleskorn — APPROVED** [review]
Thanks @FUSAKLA!

**Comment by FUSAKLA** [comment]
Rebased, not sure who to ask for approval?

**Review by charleskorn — APPROVED** [review]

**Pull request merged**